src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,563 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+
+namespace osp
+{
+
+#ifdef __GNUC__
+
+    inline int ctz(uint64_t a) noexcept { return __builtin_ctzll(a); }
+
+#elif defined(_MSC_VER)
+
+    #include <intrin.h>
+
+    inline int ctz(uint64_t a) noexcept
+    {
+        unsigned long int b = 0;
+        _BitScanForward64(&b, a);
+        return b;
+    }
+#elif
+
+    static_assert(false, "Missing ctz 'count trailing zeros' implementation for this compiler");
+
+#endif
+
+/**
+ * @brief Divide and round up
+ */
+template<typename NUM_T, typename DENOM_T>
+constexpr decltype(auto) div_ceil(NUM_T num, DENOM_T denom) noexcept
+{
+    return (num / denom) + (num % denom != 0);
+}
+
+template<typename INT_T>
+constexpr bool bit_test(INT_T block, int bit)
+{
+    return (block & (INT_T(1) << bit) ) != 0;
+}
+
+template<typename INT_T>
+constexpr void copy_bits(INT_T const* pSrc, INT_T* pDest, size_t bits) noexcept
+{
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = *pSrc;
+        pSrc ++;
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        INT_T const mask = (~INT_T(0)) << bits;
+
+        *pDest &= mask;
+        *pDest |= ( (~mask) & (*pSrc) );
+    }
+}
+
+template<typename INT_T>
+constexpr void set_bits(INT_T* pDest, size_t bits) noexcept
+{
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = ~INT_T(0);
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        *pDest |= ~( (~INT_T(0)) << bits );
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A bitset featuring hierarchical rows for fast iteration over set bits
+ *
+ * For example, when using 8-bit block sizes, the structure looks like this:
+ *
+ * Row 1: 00000101
+ * Row 0: 00000011 00000000 00001100 00000000 00000000
+ *
+ * Size: 40 bits, determined by Row 0
+ * Bits set: { 0, 1, 18, 19 }
+ *
+ * note: Bits are displayed in msb->lsb, but arrays are left->right
+ *
+ * * Row 1's bit 0 and bit 2 means Row 0's block 0 and block 2 is non-zero
+ * * Row 1's last 3 bits are always zero
+ * * Container supports sizes that aren't aligned with block sizes, which means
+ *   the last bits of Row 0's last block may be ignored
+ */
+template <typename BLOCK_INT_T = uint64_t>
+class HierarchicalBitset
```

Comment:
Does this container only ever deal with 2 rows?

Could it be genericized to support N number of rows, with the logic being collapsed by the compiler to the simpliest runtime logic?

Reply:
It seems my documentation is a bit misleading. It supports smc_maxRows, which is fixed at 8 at the moment. The max container size is smc_blockSize^smc_maxRows. Any common int width {8, 16, 32, or 64} to the power of 8 is already a really big number, so this isn't a problem.

Reply:
Gotcha.

I have a weak preference to see the number of rows set as a template parameter, but not a big deal either way.